include/aws/common/array_list_with_map.h

```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+#define AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/common.h>
+#include <aws/common/hash_table.h>
+
+struct aws_array_list_with_map {
+    struct aws_array_list list;
+    struct aws_hash_table map; /* map from the element to the index in the array */
+};
+
+AWS_EXTERN_C_BEGIN
+
+/* TODO: CBMC verification. */
+/**
+ * Initialize the array_list with map structure, which support constant time of insert, remove and get random element
+ * from the data structure.
+ *
+ * The underlying hash map will use hash_fn to compute the hash of each element. equals_fn to compute equality of two
+ * keys.
+ *
+ * @param list_with_map             Pointer of structure to initialize with
+ * @param allocator                 Allocator
+ * @param hash_fn                   Compute the hash of each element
+ * @param equals_fn                 Compute equality of two keys
+ * @param destroy_element_fn        Called when the element is removed
+ * @param item_size                 Size of each element
+ * @param initial_item_allocation   The initial number of item to allocate.
```

Comment:
item_allocation sounds kind of weird can we just call this size? From the perspective of the user this is just a set with an initial size, they don't need to care that size isn't the actual size of the list. aws_hash_table size doesn't correspond to the underlying array size either: https://github.com/awslabs/aws-c-common/blob/d9031cf4f02fe4d8c24a53bb6c8df67ce8be9f0c/source/hash_table.c#L233

Reply:
I don't see anything weird here. We allocated items for the array list. The hash table doesn't allocate anything related to the element. It stores the pointer. Those are helper as implementation details.

Reply:
Ah I see the name comes from aws_array_list_init_dynamic. I still think size makes more sense since the functionality is that of a set which is more similar to a map than a list so we should use names similar to that of the hashtable, but it's not a huge deal one way or the other.